lib/vmpooler/pool_manager.rb

```diff
@@ -20,39 +20,62 @@ def initialize(config, logger, redis, metrics)
     end
 
     # Check the state of a VM
-    def check_pending_vm(vm, pool, timeout)
+    def check_pending_vm(vm, pool, timeout, vsphere)
       Thread.new do
-        _check_pending_vm(vm, pool, timeout)
+        _check_pending_vm(vm, pool, timeout, vsphere)
       end
     end
 
-    def _check_pending_vm(vm, pool, timeout)
-      host = $vsphere[pool].find_vm(vm)
-
-      if host
+    def open_socket(host, domain=nil, timeout=5, port=22, &block)
+      Timeout.timeout(timeout) do
+        target_host = host
+        target_host = "#{host}.#{domain}" if domain
+        sock = TCPSocket.new target_host, port
         begin
-          Timeout.timeout(5) do
-            TCPSocket.new vm, 22
-          end
-          move_pending_vm_to_ready(vm, pool, host)
-        rescue
-          fail_pending_vm(vm, pool, timeout)
+          yield sock if block_given?
+        ensure
+          sock.close
+          return true
         end
-      else
-        fail_pending_vm(vm, pool, timeout)
       end
+    rescue SocketError
+    rescue Timeout::Error
     end
 
-    def fail_pending_vm(vm, pool, timeout)
-      clone_stamp = $redis.hget('vmpooler__vm__' + vm, 'clone')
+    def _check_pending_vm(vm, pool, timeout, vsphere)
+      host = vsphere.find_vm(vm)
 
-      if (clone_stamp) &&
-          (((Time.now - Time.parse(clone_stamp)) / 60) > timeout)
+      if ! host
+        fail_pending_vm(vm, pool, timeout, false)
+        return
+      end
+      socket = open_socket vm
+      fail_pending_vm(vm, pool, timeout) if ! socket
+      move_pending_vm_to_ready(vm, pool, host)
+    rescue
+      fail_pending_vm(vm, pool, timeout)
```

Comment:
for example this could just be:

```ruby
  open_socket vm
  move_pending_vm_to_ready(vm, pool, host)
rescue
  fail_pending_vm(vm, pool, timeout)
```